contrib/cshar/makekit/makekit.1
.TH MAKEKIT 1
.\" $Header: /bsdi/MASTER/BSDI_OS/contrib/cshar/makekit/makekit.1,v 1.2 1993/03/18 22:11:56 polk Exp $
.SH NAME
makekit \- split files up into shell archive packages
.SH SYNOPSIS
.B makekit
[
.B -1
] [
.B -e
] [
.B -x
] [
.BI -h\| #
] [
.BI -i\| name
] [
.BI -k\| #
] [
.B -m
] [
.BI -n\| name
] [
.BI -o\| name
] [
.B -p
] [
.BI -s\| #[k]
] [
.BI -t\| text
] [ file... ]
.SH DESCRIPTION
.I Makekit
reads a list of files and directories, determines their sizes,
and parcels them up into a series of shell archives such that all the
archives are of reasonable size.
It then invokes
.IR cshar (1l)
to actually create the archives.
.PP
By default, no archive will be larger than about 50,000 bytes; this may be
changed by using the ``\-s'' option.
If the number given with the ``\-s'' option ends with the letter ``k''
then the size is multiplied by 1024, otherwise it is taken to be the
desired maximum size, in bytes.
Each archive will have a name that looks like
.IR Part nn,
where ``nn'' represents the two-digit sequence number (with leading zero
if needed).
The leader part of the archive name may be changed with the ``\-n'' option.
The ``\-n'' is also useful when write permission to the directory being
archive is denied; e.g., ``\-n/tmp/KERNEL.''
.PP
.I Makekit
reads its list of files on the command line, or standard input
if none are given.
It is also possible to specify an input filename with the ``\-i'' option.
The input should contain a list of files, one to a line, to separate.
In addition, if each input line looks like this:
.RS
filename\ \ \ whitespace\ \ \ optional-digits\ \ \ whitespace\ \ \ text
.RE
then
.I makekit
will ignore the spaces and digits, but remember the text associated with
each file, and output it with the filename when generating the ``shipping
manifest.''
Further, the ``\-h'' option may be given to have the program skip the
indicated number of lines in the input; this option is provided so that
.I makekit
can more easily re-parse the manifests it has generated.
.PP
The generated manifest will be sent to the standard output.
An alternate output file may be given by using the ``\-o'' option; if
the output file exists,
.I makekit
will try to rename it with an extension of
.IR \&.BAK \&.
If the ``\-o'' option is used,
.I makekit
will add that name to the list of files to be archived; the ``\-e''
option may be given to exclude the manifest from the list.
.PP
The ``\-m'' option is the same as giving the options,
\&``-iMANIFEST -oMANIFEST -h2.''
This is a common way to regenerate a set of archives after the first
use of
.I makekit
in a directory.
.PP
If a large number of kits has to be generated, you may need to give
the ``\-k'' option to increase the maximum number of kits to be
generated.
.PP
After partitioning the files and directories,
.I makekit
calls
.I cshar
with the proper options to generate archives in a series.
Each resultant archive will, when executed, check to see if all the parts
are present.
If the ``\-1'' option is used, then
.I makekit
will not instruct
.I cshar
to generate the checks (by not passing on the ``\-n'' and ``\-e'' options).
By using the ``\-t'' option, you can specify a line of starting instructions
to display to the recipient when all pieces have been unpacked.
This is useful when resending part of a series that has probably already
been unpacked by the recipient.
See
.I cshar
for more information on multi-part archives.
If the ``\-x'' option is used,
.I cshar
is not called, but the manifest is still created.
.PP
.I Makekit
normally reorders its input so that the archives are as ``dense'' as
possible, with the exception that directories are given priority over
files, and a file named
.I README
is the first of all.
The manifest is also sorted in alphabetical order; this makes it easy
to locate ``missing'' files when the distribution is a large one.
The ``\-p'' option may be used to override both sortings, however,
and preserve the original order of the input list in generating
both the manifest, and the shell archives.
.SH NOTES
.I Makekit
tries to partition the files so that all directories are in the first archive.
This usually means the first archive must be the first one to be unpacked.
.PP
Saying ``the `\-k' option is to help prevent runaway packaging'' is probably
.I "post hoc propter hoc"
reasoning.
.SH "SEE ALSO"
findsrc(1l), cshar(1l)


contrib/cshar/makekit/makekit.c
/*
**  MAKEKIT
**  Split up source files into reasonably-sized shar lists.
*/
#include "shar.h"
#ifdef	RCSID
static char RCS[] =
	"$Header: /bsdi/MASTER/BSDI_OS/contrib/cshar/makekit/makekit.c,v 1.1.1.1 1993/02/23 18:00:03 polk Exp $";
#endif	/* RCSID */


/*
**  Our block of information about the files we're doing.
*/
typedef struct _block {
    char	*Name;			/* Filename			*/
    char	*Text;			/* What it is			*/
    int		 Where;			/* Where it is			*/
    int		 Type;			/* Directory or file?		*/
    long	 Bsize;			/* Size in bytes		*/
} BLOCK;


/*
**  Our block of information about the archives we're making.
*/
typedef struct _archive {
    int		 Count;			/* Number of files		*/
    long	 Asize;			/* Bytes used by archive	*/
} ARCHIVE;


/*
**  Format strings; these are strict K&R so you shouldn't have to change them.
*/
#define FORMAT1		" %-25s %2d\t%s\n"
#define FORMAT2		"%s%2.2d"
#ifdef	FMT02d
#undef FORMAT2
#define FORMAT2		"%s%02.2d"	/* I spoke too soon...		*/
#endif	/* FMT02d */


/*
**  Global variables.
*/
char	*InName;			/* File with list to pack	*/
char	*OutName;			/* Where our output goes	*/
char	*SharName = "Part";		/* Prefix for name of each shar	*/
char	*Trailer;			/* Text for shar to pack in	*/
char	*TEMP;				/* Temporary manifest file	*/
#ifdef	MSDOS
char	*FLST;				/* File with list for shar	*/
#endif	/* MSDOS */
int	 ArkCount = 20;			/* Max number of archives	*/
int	 ExcludeIt;			/* Leave out the output file?	*/
int	 Header;			/* Lines of prolog in input	*/
int	 Preserve;			/* Preserve order for Manifest?	*/
int	 Working = TRUE;		/* Call shar when done?		*/
long	 Size = 55000;			/* Largest legal archive size	*/


/*
**  Sorting predicate to put README or MANIFEST first, then directories,
**  then larger files, then smaller files, which is how we want to pack
**  stuff in archives.
*/
static int
SizeP(t1, t2)
    BLOCK	*t1;
    BLOCK	*t2;
{
    long	 i;

    if (t1->Type == F_DIR)
	return(t2->Type == F_DIR ? 0 : -1);
    if (t2->Type == F_DIR)
	return(1);
    if (EQn(t1->Name, "README", 6) || (OutName && EQ(t1->Name, OutName)))
	return(-1);
    if (EQn(t2->Name, "README", 6) || (OutName && EQ(t2->Name, OutName)))
	return(1);
    return((i = t1->Bsize - t2->Bsize) == 0L ? 0 : (i < 0L ? -1 : 1));
}


/*
**  Sorting predicate to get things in alphabetical order, which is how
**  we write the Manifest file.
*/
static int
NameP(t1, t2)
    BLOCK	*t1;
    BLOCK	*t2;
{
    int		 i;

    return((i = *t1->Name - *t2->Name) ? i : strcmp(t1->Name, t2->Name));
}


/*
**  Skip whitespace.
*/
static char *
Skip(p)
    REGISTER char	*p;
{
    while (*p && WHITE(*p))
	p++;
    return(p);
}


/*
**  Signal handler.  Clean up and die.
*/
static sigret_t
Catch(s)
    int		 s;
{
    int		 e;

    e = errno;
    if (TEMP)
	(void)unlink(TEMP);
#ifdef	MSDOS
    if (FLST)
	(void)unlink(FLST);
#endif	/* MSDOS */
    Fprintf(stderr, "Got signal %d, %s.\n", s, Ermsg(e));
    exit(1);
}


main(ac, av)
    REGISTER int	 ac;
    char		*av[];
{
    REGISTER FILE	*F;
    REGISTER FILE	*In;
    REGISTER BLOCK	*t;
    REGISTER ARCHIVE	*k;
    REGISTER char	*p;
    REGISTER int	 i;
    REGISTER int	 lines;
    REGISTER int	 Value;
    BLOCK		*Table;
    BLOCK		*TabEnd;
    ARCHIVE		*Ark;
    ARCHIVE		*ArkEnd;
    char		 buff[BUFSIZ];
    long		 lsize;
    int			 LastOne;
    int			 Start;
    int			 Notkits;
    char		 EndArkNum[20];
    char		 CurArkNum[20];

    /* Collect input. */
    Value = FALSE;
    Notkits = FALSE;
    while ((i = getopt(ac, av, "1eh:i:k:n:mo:ps:t:x")) != EOF)
	switch (i) {
	default:
	    Fprintf(stderr, "usage: makekit %s\n        %s\n",
		    "[-1] [-e] [-x] [-k #] [-s #[k]] [-n Name] [-t Text]",
		    "[-p] [-m | -i MANIFEST -o MANIFEST -h 2] [file ...]");
	    exit(1);
	case '1':
	    Notkits = TRUE;
	    break;
	case 'e':
	    ExcludeIt = TRUE;
	    break;
	case 'h':
	    Header = atoi(optarg);
	    break;
	case 'i':
	    InName = optarg;
	    break;
	case 'k':
	    ArkCount = atoi(optarg);
	    break;
	case 'm':
	    InName = OutName = "MANIFEST";
	    Header = 2;
	    break;
	case 'n':
	    SharName = optarg;
	    break;
	case 'o':
	    OutName = optarg;
	    break;
	case 'p':
	    Preserve = TRUE;
	    break;
	case 's':
	    Size = atoi(optarg);
	    if (IDX(optarg, 'k') || IDX(optarg, 'K'))
		Size *= 1024;
	    break;
	case 't':
	    Trailer = optarg;
	    break;
	case 'x':
	    Working = FALSE;
	    break;
	}
    ac -= optind;
    av += optind;

    /* Write the file list to a temp file. */
    TEMP = mktemp("/tmp/maniXXXXXX");
    F = fopen(TEMP, "w");
    SetSigs(TRUE, Catch);
    if (av[0])
	/* Got the arguments on the command line. */
	while (*av)
	    Fprintf(F, "%s\n", *av++);
    else {
	/* Got the name of the file from the command line. */
	if (InName == NULL)
	    In = stdin;
	else if ((In = fopen(InName, "r")) == NULL) {
	    Fprintf(stderr, "Can't read %s as manifest, %s.\n",
		    InName, Ermsg(errno));
	    exit(1);
	}
	/* Skip any possible prolog, then output rest of file. */
	while (--Header >= 0 && fgets(buff, sizeof buff, In))
	    ;
	if (feof(In)) {
	    Fprintf(stderr, "Nothing but header lines in list!?\n");
	    exit(1);
	}
	while (fgets(buff, sizeof buff, In))
	    fputs(buff, F);
	if (In != stdin)
	    (void)fclose(In);
    }
    (void)fclose(F);

    /* Count number of files, allow for NULL and our output file. */
    F = fopen(TEMP, "r");
    for (lines = 2; fgets(buff, sizeof buff, F); lines++)
	;
    rewind(F);

    /* Read lines and parse lines, see if we found our OutFile. */
    Table = NEW(BLOCK, lines);
    for (t = Table, Value = FALSE, lines = 0; fgets(buff, sizeof buff, F); ) {
	/* Read line, skip first word, check for blank line. */
	if (p = IDX(buff, '\n'))
	    *p = '\0';
	else
	    Fprintf(stderr, "Warning, line truncated:\n%s\n", buff);
	p = Skip(buff);
	if (*p == '\0')
	    continue;

	/* Copy the line, snip off the first word. */
	for (p = t->Name = COPY(p); *p && !WHITE(*p); p++)
	    ;
	if (*p)
	    *p++ = '\0';

	/* Skip <spaces><digits><spaces>; remainder is the file description. */
	for (p = Skip(p); *p && isascii(*p) && isdigit(*p); )
	    p++;
	t->Text = Skip(p);

	/* Get file type. */
	if (!GetStat(t->Name)) {
	    Fprintf(stderr, "Can't stat %s (%s), skipping.\n",
		    t->Name, Ermsg(errno));
	    continue;
	}
	t->Type = Ftype(t->Name);

	/* Guesstimate its size when archived:  prolog, plus one char/line. */
	t->Bsize = strlen(t->Name) * 3 + 200;
	if (t->Type == F_FILE) {
	    lsize = Fsize(t->Name);
	    t->Bsize += lsize + lsize / 60;
	}
	if (t->Bsize > Size) {
	    Fprintf(stderr, "At %ld bytes, %s is too big for any archive!\n",
		    t->Bsize, t->Name);
	    exit(1);
	}

	/* Is our ouput file there? */
	if (!Value && OutName && EQ(OutName, t->Name))
	    Value = TRUE;

	/* All done -- advance to next entry. */
	t++;
    }
    (void)fclose(F);
    (void)unlink(TEMP);
    SetSigs(S_RESET, (sigret_t (*)())NULL);

    /* Add our output file? */
    if (!ExcludeIt && !Value && OutName) {
	t->Name = OutName;
	t->Text = "This shipping list";
	t->Type = F_FILE;
	t->Bsize = lines * 60;
	t++;
    }

    /* Sort by size, get archive space. */
    lines = t - Table;
    TabEnd = &Table[lines];
    if (!Preserve)
	qsort((char *)Table, lines, sizeof Table[0], SizeP);
    Ark = NEW(ARCHIVE, ArkCount);
    ArkEnd = &Ark[ArkCount];

    /* Loop through the pieces, and put everyone into an archive. */
    for (t = Table; t < TabEnd; t++) {
	for (k = Ark; k < ArkEnd; k++)
	    if (t->Bsize + k->Asize < Size) {
		k->Asize += t->Bsize;
		t->Where = k - Ark;
		k->Count++;
		break;
	    }
	if (k == ArkEnd) {
	    Fprintf(stderr, "'%s' doesn't fit -- need more then %d archives.\n",
		    t->Name, ArkCount);
	    exit(1);
	}
	/* Since our share doesn't build sub-directories... */
	if (t->Type == F_DIR && k != Ark)
	    Fprintf(stderr, "Warning:  directory '%s' is in archive %d\n",
		    t->Name, k - Ark + 1);
    }

    /* Open the output file. */
    if (OutName == NULL)
	F = stdout;
    else {
	if (GetStat(OutName)) {
#ifdef	BACKUP_PREFIX
	    (void)sprintf(buff, "%s%s", BACKUP_PREFIX, OutName);
#else
	    /* Handle /foo/bar/VeryLongFileName.BAK for non-BSD sites. */
	    (void)sprintf(buff, "%s.BAK", OutName);
	    p = (p = RDX(buff, '/')) ? p + 1 : buff;
	    if (strlen(p) > 14)
		/* ... well, sort of handle it. */
		(void)strcpy(&p[10], ".BAK");
#endif	/* BACKUP_PREFIX */
	    Fprintf(stderr, "Renaming %s to %s\n", OutName, buff);
	    (void)rename(OutName, buff);
	}
	if ((F = fopen(OutName, "w")) == NULL) {
	    Fprintf(stderr, "Can't open '%s' for output, %s.\n",
		    OutName, Ermsg(errno));
	    exit(1);
	}
    }

    /* Sort the shipping list, then write it. */
    if (!Preserve)
	qsort((char *)Table, lines, sizeof Table[0], NameP);
    Fprintf(F, "   File Name\t\tArchive #\tDescription\n");
    Fprintf(F, "-----------------------------------------------------------\n");
    for (t = Table; t < TabEnd; t++)
	Fprintf(F, FORMAT1, t->Name, t->Where + 1, t->Text);

    /* Close output.  Are we done? */
    if (F != stdout)
	(void)fclose(F);
    if (!Working)
	exit(0);

    /* Find last archive number. */
    for (i = 0, t = Table; t < TabEnd; t++)
	if (i < t->Where)
	    i = t->Where;
    LastOne = i + 1;

    /* Find archive with most files in it. */
    for (i = 0, k = Ark; k < ArkEnd; k++)
	if (i < k->Count)
	    i = k->Count;

    /* Build the fixed part of the argument vector. */
    av = NEW(char*, i + 10);
    av[0] = "cshar";
    i = 1;
    if (Trailer) {
	av[i++] = "-t";
	av[i++] = Trailer;
    }
    if (Notkits == FALSE) {
	(void)sprintf(EndArkNum, "%d", LastOne);
	av[i++] = "-e";
	av[i++] = EndArkNum;
	av[i++] = "-n";
	av[i++] = CurArkNum;
    }
#ifdef	MSDOS
    av[i++] = "-i";
    av[i++] = FLST = mktemp("/tmp/manlXXXXXX");
#endif	/* MSDOS */

    av[i++] = "-o";
    av[i++] = buff;

    /* Call shar to package up each archive. */
    for (Start = i, i = 0; i < LastOne; i++) {
	(void)sprintf(CurArkNum, "%d", i + 1);
	(void)sprintf(buff, FORMAT2, SharName, i + 1);
#ifndef	MSDOS
	for (lines = Start, t = Table; t < TabEnd; t++)
	    if (t->Where == i)
		av[lines++] = t->Name;
	av[lines] = NULL;
#else
	if ((F = fopen(FLST, "w")) == NULL) {
	    Fprintf(stderr, "Can't open list file '%s' for output, %s.\n",
		    FLST, Ermsg(errno));
	    exit(1);
	}
	for (t = Table; t < TabEnd; t++)
	    if (t->Where == i)
		Fprintf(F, "%s\n", t->Name);
	(void)fclose(F);
#endif /* MSDOS */
	Fprintf(stderr, "Packing kit %d...\n", i + 1);
	if (lines = Execute(av))
	    Fprintf(stderr, "Warning:  shar returned status %d.\n", lines);
    }

#ifdef	MSDOS
    (void)unlink(FLST);
#endif	/* MSDOS */
    /* That's all she wrote. */
    exit(0);
}
